src/uct/ib/dc/dc_mlx5.c

```diff
@@ -771,71 +792,123 @@ static void uct_dc_mlx5_iface_dcis_destroy(uct_dc_mlx5_iface_t *iface,
         }
         uct_ib_mlx5_qp_mmio_cleanup(&iface->tx.dcis[dci_index].txwq.super,
                                     iface->tx.dcis[dci_index].txwq.reg);
+        iface->tx.dcis[dci_index].initialized = 0;
     }
 
     for (pool_index = 0; pool_index < num_dci_pools; pool_index++) {
         uct_dc_mlx5_iface_dci_pool_destroy(&iface->tx.dci_pool[pool_index]);
     }
 
+    iface->tx.dci_counter = 0;
     ucs_free(iface->tx.dcis);
 }
 
+ucs_status_t
+uct_dc_mlx5_iface_create_dci_pool(uct_dc_mlx5_iface_t *iface,
+                                  const uct_dc_mlx5_dci_config_t *config,
+                                  uint8_t pool_size, uint8_t *pool_index_p)
+{
+    const int pool_index = iface->tx.num_dci_pools;
+    uct_dc_mlx5_dci_pool_t *dci_pool;
+
+    ucs_assert(iface->tx.num_dci_pools < UCT_DC_MLX5_IFACE_MAX_DCI_POOLS);
+    ucs_debug("creating dci pool %u with %u QPs", pool_index, iface->tx.ndci);
+    dci_pool        = &iface->tx.dci_pool[pool_index];
+    dci_pool->stack = ucs_calloc(pool_size, sizeof(*dci_pool->stack),
+                                 "dci pool stack");
+    if (dci_pool->stack == NULL) {
+        ucs_error("failed to allocate dci_pool stack");
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    ucs_arbiter_init(&dci_pool->arbiter);
+    dci_pool->stack_top         = 0;
+    dci_pool->release_stack_top = -1;
+    dci_pool->capacity          = pool_size;
+    dci_pool->size              = 0;
+    dci_pool->config_key        = config->u64;
+
+    iface->tx.num_dci_pools++;
+    *pool_index_p = pool_index;
+
+    return UCS_OK;
+}
+
+ucs_status_t
+uct_dc_mlx5_dci_pool_get_or_create(uct_dc_mlx5_iface_t *iface,
+                                   const uct_dc_mlx5_dci_config_t *dci_config,
+                                   uint8_t *pool_index_p)
+{
+    khiter_t it;
+    ucs_kh_put_t kh_put_ret;
+    ucs_status_t status;
+    uint8_t pool_size;
+
+    it = kh_get(uct_dc_mlx5_config_hash, &iface->dc_config_hash,
+                dci_config->u64);
+    if (it != kh_end(&iface->dc_config_hash)) {
+        *pool_index_p = kh_value(&iface->dc_config_hash, it);
+        return UCS_OK;
+    }
+
+    it = kh_put(uct_dc_mlx5_config_hash, &iface->dc_config_hash,
+                dci_config->u64, &kh_put_ret);
+    if (kh_put_ret == UCS_KH_PUT_FAILED) {
+        ucs_error("failed to kh_put a new dci configuration");
+        return UCS_ERR_EXCEEDS_LIMIT;
+    }
+
+    pool_size = (dci_config->key.flags & UCT_DC_MLX5_DCI_CONFIG_KEEPALIVE) ?
```

Comment:
1. can uct_dc_mlx5_iface_create_dci_pool calc pool_size internally?
2. as a preparation task, we can remove keepalive from DC since it's not useful there anyway @brminich FYI